File: api/lib/flirtual_web/components/input/duel_range_slider.ex
defmodule FlirtualWeb.Components.Input.DuelRangeSlider do
  use Phoenix.Component

  import FlirtualWeb.Utilities

  defp range_input(assigns) do
    ~H"""
    <input {attr_merge(assigns, %{
      type: "range",
      class: ["absolute w-full bg-transparent appearance-none pointer-events-none range-thumb:bg-brand-gradient range-thumb:pointer-events-auto range-thumb:border-none range-thumb:w-6 range-thumb:h-6 range-thumb:shadow-brand-1 range-thumb:rounded-full"]
    })}>
    """
  end

  def duel_range_slider(assigns) do
    assigns =
      assigns
      |> assign_new(:step, fn -> 1 end)

    min = assigns.min |> List.first()
    min_default = min[:default] || 0
    min_limit = min[:limit] || 0

    max = assigns.max |> List.first()
    max_default = max[:default] || 100
    max_limit = max[:limit] || 100

    ~H"""
    <div x-component="duel_range_slider" x-bind="xRoot" {attr_merge(assigns, %{
      class: ["relative flex items-center h-6"],
      "data-min-default": min_default,
      "data-max-default": max_default,
      "data-min-limit": min_limit,
      "data-max-limit": max_limit
    }, [:min, :max])}>
      <div class="absolute w-full h-2 rounded-full bg-brand-black shadow-brand-1"/>
      <div x-bind="xSelection" class="absolute h-2 rounded-full bg-brand-gradient"/>
      <.range_input x-bind="xLowerInput"/>
      <.range_input x-bind="xUpperInput"/>
    </div>
    """
  end
end

File: api/assets/js/components/duel_range_slider.ts
import { component } from "../alpine-component";
import { clamp } from "../utilities";

export default component(function () {
	return {
		_min: null,
		get min() {
			const { $data } = this;
			return $data._min ?? $data.default.min;
		},
		set min(value: number) {
			const { $data } = this;
			this.$data._min = clamp(value, $data.limit.min, $data.max);
		},

		_max: null,
		get max() {
			const { $data } = this;
			return $data._max ?? $data.default.max;
		},
		set max(value: number) {
			const { $data } = this;
			this.$data._max = clamp(value, $data.min, $data.limit.max);
		},

		get default() {
			const { $data, $el } = this;

			return {
				min: Number.parseInt($el.getAttribute("data-min-default") as string) || $data.limit.min,
				max: Number.parseInt($el.getAttribute("data-max-default") as string) || $data.limit.max
			};
		},

		get step() {
			return Number.parseInt(this.$el.getAttribute("step") as string);
		},
		get limit() {
			const min = Number.parseInt(this.$el.getAttribute("data-min-limit") as string);
			const max = Number.parseInt(this.$el.getAttribute("data-max-limit") as string);

			return {
				min,
				max,
				diff: max - min
			};
		},

		xRoot: {},
		xLowerInput: {
			":value"() {
				return this.$data.min;
			},
			":min"() {
				return this.$data.limit.min;
			},
			":max"() {
				return this.$data.limit.max;
			},
			":step"() {
				return this.$data.step;
			},
			"@input"(event: InputEvent) {
				if (!(event.target instanceof HTMLInputElement)) return;
				this.$data.min = event.target.valueAsNumber;
			}
		},
		xUpperInput: {
			":value"() {
				return this.$data.max;
			},
			":min"() {
				return this.$data.limit.min;
			},
			":max"() {
				return this.$data.limit.max;
			},
			":step"() {
				return this.$data.step;
			},
			"@input"(event: InputEvent) {
				if (!(event.target instanceof HTMLInputElement)) return;
				this.$data.max = event.target.valueAsNumber;
			}
		},
		xSelection: {
			":style"() {
				const {
					$data: { min, max, limit }
				} = this;

				return {
					"margin-left": `${((min - limit.min) / limit.diff) * 100}%`,
					width: `${((max - min) / limit.diff) * 100}%`
				};
			}
		}
	};
});
